sources/pciclient/src/lib.rs

```diff
@@ -0,0 +1,488 @@
+    process::Command,
+};
+
+use error::{
+    CommandFailureSnafu, ExecutionFailureSnafu, MissingRequiredFieldSnafu,
+    ParseListDevicesOutputFailureSnafu,
+};
+use snafu::{ensure, OptionExt, ResultExt};
+
+const AMAZON_VENDOR_CODE: &str = "1d0f";
+const NETWORK_ETHERNET_CONTROLLER_CLASS: &str = "0200";
+const EFA_KEYWORD: &str = "efa";
+
+const LSPCI_PATH: &str = "/usr/bin/lspci";
+
+/// This [`ListDevicesParam`] is based on the "-d" flag for `lspci`, which
+/// allows selection/filtering of the output. Spec for "-d":
+/// ```
+/// -d [<vendor>]:[<device>][:<class>[:<prog-if>]]
+///
+/// Show only devices with specified vendor, device, class ID,
+/// and programming interface.  The ID's are given in
+/// hexadecimal and may be omitted or given as "*", both
+/// meaning "any value". The class ID can contain "x"
+/// characters which stand for "any digit".
+/// ```
+#[derive(Debug, Default, PartialEq)]
+pub struct ListDevicesParam {
+    vendor: String,
+    device: String,
+    class: String,
+    program_interface: String,
+}
+
+impl ListDevicesParam {
+    /// Convert the ListDevicesParam into proper command line arguments so that
+    /// we can feed it to the lspci() function where it makes the actual call to the
+    /// lspci binary.
+    ///
+    fn to_command_args(self) -> Vec<OsString> {
+        let mut args: Vec<String> = vec!["-n".into(), "-m".into()];
+        if self != ListDevicesParam::default() {
+            let parts = vec![self.vendor, self.device, self.class, self.program_interface];
+            let additional_args = parts.join(":");
+            args.push("-d".to_string());
+            args.push(additional_args);
+        }
+        args.into_iter().map(OsString::from).collect()
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct ListDevicesParamBuilder {
+    param: ListDevicesParam,
+}
+
+impl ListDevicesParamBuilder {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn vendor(mut self, vendor: impl Into<String>) -> Self {
+        self.param.vendor = vendor.into();
+        self
+    }
+
+    pub fn device(mut self, device: impl Into<String>) -> Self {
+        self.param.device = device.into();
+        self
+    }
+
+    pub fn class(mut self, class: impl Into<String>) -> Self {
+        self.param.class = class.into();
+        self
+    }
+
+    pub fn program_interface(mut self, program_interface: impl Into<String>) -> Self {
+        self.param.program_interface = program_interface.into();
+        self
+    }
+
+    pub fn build(self) -> ListDevicesParam {
+        self.param
+    }
+}
+
+/// The [`ListDevicesOutput`] is based on the output format that is decorated by "-n -m".
+#[derive(Debug, Default, PartialEq, Eq)]
+pub struct ListDevicesOutput {
+    pci_slot: String,
+    class: String,
+    vendor: String,
+    device: String,
+    subsystem_vendor: Option<String>,
+    subsystem_device: Option<String>,
+    revision: Option<String>,
+    program_interface: Option<String>,
```

Comment:
These are private, so we probably need to provide a way to get some of them? Or is the expectation to add functions to this library to do anything with this output?

Reply:
A high-level API is a good idea if you have an idea for one, it makes it easier to change the underlying implementation if we need to.

I think you have a case now where you need to choose how to expose the results -- I think `list_devices`' output is all private with no way to access the results.

Reply:
Right, given that list_devices already emits the output, it is better to add accessor to the outputs or make them public.